src/gc.rs

```diff
@@ -39,41 +39,99 @@ use crate::GC_ALLOCATOR;
 /// free `Gc` values behind the scenes.
 #[derive(PartialEq, Eq, Debug)]
 pub struct Gc<T: ?Sized> {
-    pub(crate) objptr: NonNull<GcBox<T>>,
+    pub(crate) ptr: NonNull<GcBox<T>>,
+    _phantom: PhantomData<T>,
 }
 
 impl<T> Gc<T> {
     /// Constructs a new `Gc<T>`.
     pub fn new(v: T) -> Self {
         Gc {
-            objptr: unsafe { NonNull::new_unchecked(GcBox::new(v)) },
+            ptr: unsafe { NonNull::new_unchecked(GcBox::new(v)) },
+            _phantom: PhantomData,
         }
     }
+
+    /// Constructs a new `Gc<MaybeUninit<T>>` which is capable of storing data
+    /// up-to the size permissible by `layout`.
+    ///
+    /// This can be useful if you want to store a value with a custom layout,
+    /// but have the collector treat the value as if it were T.
+    ///
+    /// If `layout` is insufficiently sized to store `T`, a Gc value will not be
+    /// allocated. This is because it's possible to accidentally access out-of-
+    /// bounds memory when treating the Gc value as if it contains a type larger
+    /// than its actual size.
+    pub fn new_from_layout(layout: Layout) -> Option<Gc<MaybeUninit<T>>> {
+        if layout.size() < Layout::new::<T>().size() {
+            return None;
+        }
+        Some(Gc::from_inner(GcBox::new_from_layout(layout)))
+    }
 }
 
 impl Gc<dyn Any> {
     pub fn downcast<T: Any>(&self) -> Result<Gc<T>, Gc<dyn Any>> {
         if (*self).is::<T>() {
-            let ptr = self.objptr.cast::<GcBox<T>>();
+            let ptr = self.ptr.cast::<GcBox<T>>();
             forget(self);
-            Ok(Gc { objptr: ptr })
+            Ok(Gc::from_inner(ptr))
         } else {
-            Err(Gc {
-                objptr: self.objptr,
-            })
+            Err(Gc::from_inner(self.ptr))
         }
     }
 }
 
 impl<T: ?Sized> Gc<T> {
     /// Get a raw pointer to the underlying value `T`.
     pub fn into_raw(this: Self) -> *const T {
-        let ptr: *const T = &*this;
-        ptr
+        this.ptr.as_ptr() as *const T
     }
 
     pub fn ptr_eq(this: &Self, other: &Self) -> bool {
-        this.objptr.as_ptr() == other.objptr.as_ptr()
+        this.ptr.as_ptr() == other.ptr.as_ptr()
+    }
+
+    pub fn from_raw(raw: *const T) -> Gc<T> {
+        Gc {
+            ptr: unsafe { NonNull::new_unchecked(raw as *mut GcBox<T>) },
+            _phantom: PhantomData,
+        }
+    }
+
+    fn from_inner(ptr: NonNull<GcBox<T>>) -> Self {
+        Self {
+            ptr,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T> Gc<MaybeUninit<T>> {
+    /// Be aware that unlike with other smart-pointers such as
```

Comment:
Result of in person discussion: we'll probably remove this doc comment and replace it with something like "see [maybeuninit::assume_init]".

Reply:
Fixed: 0024ec1b173a265f1a780822563fd2f823c91ab3